src/main/go/RelpConnection.go

```diff
@@ -0,0 +1,293 @@
+func (relpConn *RelpConnection) Disconnect() bool {
+	if relpConn.state != STATE_OPEN {
+		panic("Cannot disconnect, connection was not OPEN")
+	}
+	relpRequest := RelpFrameTX{RelpFrame{
+		transactionId: relpConn.txId,
+		cmd:           RELP_CLOSE,
+		dataLength:    0,
+		data:          nil,
+	}}
+
+	closerBatch := RelpBatch{}
+	closerBatch.Init()
+
+	reqId := closerBatch.PutRequest(&relpRequest)
+	err := relpConn.SendBatch(&closerBatch)
+	success := false
+	closeResp, err := closerBatch.GetResponse(reqId)
+	if err == nil && closeResp != nil && closeResp.dataLength == 0 {
+		success = true
+	}
+
+	if success {
+		// if sending CLOSE command was successful, close connection and set state to CLOSED
+		relpConn.TearDown()
+	}
+
+	return success
+}
+
+// Commit commits the RELP batch to the server
+func (relpConn *RelpConnection) Commit(batch *RelpBatch) error {
+	if relpConn.state != STATE_OPEN {
+		panic("Can't commit, connection was in state other than OPEN.")
+	}
+
+	relpConn.state = STATE_COMMIT
+	err := relpConn.SendBatch(batch)
+	relpConn.state = STATE_OPEN
+	return err
+}
+
+// SendBatch sends the RELP frames to the server in the given batch.
+// The frames are sent asynchronously, and the server ACKs are checked after sending.
+func (relpConn *RelpConnection) SendBatch(batch *RelpBatch) error {
+	log.Printf("SendBatch.Entry> Work queue length: %v, Window.pending: %v : %v\n",
+		batch.GetWorkQueueLen(), len(relpConn.window.pending), relpConn.window.pending)
+	// send a batch of requests
+	for batch.GetWorkQueueLen() > 0 {
+		reqId := batch.PopWorkQueue()
+		relpRequest, err := batch.GetRequest(reqId)
+		if err != nil {
+			log.Fatalln("Could not get request from batch")
+		}
+
+		// relp Request-Response txId
+		// <txId is here> <command> <len> <data> NL
+		// make sure txId loops 1 - 999 999 999
+		if relpConn.txId >= 999_999_999 {
+			relpConn.txId = 1
+		} else {
+			relpConn.txId++
+		}
+		relpRequest.transactionId = relpConn.txId
+		log.Printf("SendBatch> Sending request\n%v %v %v '%v'\nfrom batch\n", relpRequest.transactionId, relpRequest.cmd,
+			relpRequest.dataLength, string(relpRequest.data))
+
+		relpConn.window.PutPending(relpConn.txId, reqId)
+		log.Println("SendBatch> Put pending: ", relpConn.txId, reqId)
+
+		go func() {
+			sendErr := relpConn.SendRelpRequest(relpRequest)
+			if sendErr != nil {
+				log.Printf("Error sending relp request: '%v'\n", err.Error())
+			}
+		}()
+	}
+
+	ackErr := relpConn.ReadAcks(batch)
+	if ackErr != nil {
+		// ACK timeout or other failure
+		return ackErr
+	}
+	return nil
+}
+
+// ReadAcks reads the ACKs from the given batch.
+func (relpConn *RelpConnection) ReadAcks(batch *RelpBatch) error {
+	log.Printf("ReadAcks.Entry> Reading ACKs for batchID: %v\n", batch.requestId)
+	var parser *RelpParser = nil
+	notComplete := relpConn.window.Size() > 0
+	var cn = *relpConn.connection
+
+	for notComplete { // until window is empty
+		tmp := make([]byte, 1)
```

Comment:
Larger buffer size would allow better performance. rlp_01 uses preAllocRxBuffer for this

Reply:
changed to use pre-allocated buffers in relpConnection struct
commit 4c66d9b